components/hab/src/scaffolding.rs

```diff
@@ -0,0 +1,167 @@
+                "go-scaffolding" => {
+                    let ident = PackageIdent::from_str("core/scaffolding-go")?;
+                    let _ = ui.status(Status::Using, &format!("go scaffolding '{}'", ident));
+                    Ok(Some(ident))
+                }
+                "core/scaffolding-node" |
+                "node" |
+                "node.js" |
+                "javascript" |
+                "js" => {
+                    let ident = PackageIdent::from_str("core/scaffolding-node")?;
+                    let _ = ui.status(Status::Using, &format!("node scaffolding '{}'", ident));
+                    Ok(Some(ident))
+                }
+                _ => {
+                    let ident = PackageIdent::from_str(scaffold)?;
+                    let _ = ui.status(Status::Using, &format!("custom scaffolding: '{}'", ident));
+                    Ok(Some(ident))
+                }
+            }
+        }
+        // If nothing was passed try to autodiscover the codebase
+        // I'm not saying it's magic. But, MAGIC.
+        None => magic_function(ui),
+    }
+}
+fn magic_function(ui: &mut UI) -> Result<Option<PackageIdent>> {
+    // Determine if the current dir has an app that can use
+    // one of our scaffoldings and use it by default
+    try!(ui.begin("Attempting autodiscovery "));
+    try!(ui.para(
+        "No scaffolding type was provided. Let's see if we can figure out \
+        what kind of application you're planning to package.",
+    ));
+    let current_path = Path::new(".");
+    if is_project_ruby(&current_path) {
+        let ident = PackageIdent::from_str("core/scaffolding-ruby")?;
+        try!(ui.begin("We've detected your app as Ruby"));
+        let _ = ui.status(Status::Using, &format!("scaffolding package: '{}'", ident));
+        try!(ui.para(""));
+        Ok(Some(ident))
+    } else if is_project_golang(&current_path) {
+        let ident = PackageIdent::from_str("core/scaffolding-go")?;
+        try!(ui.begin("We've detected your app as Golang"));
+        let _ = ui.status(Status::Using, &format!("scaffolding package: '{}'", ident));
+        try!(ui.para(""));
+        Ok(Some(ident))
+    } else if is_project_node(&current_path) {
+        let ident = PackageIdent::from_str("core/scaffolding-node")?;
+        try!(ui.begin("We've detected your app as Node.js"));
+        let _ = ui.status(Status::Using, &format!("scaffolding package: '{}'", ident));
+        try!(ui.para(""));
+        Ok(Some(ident))
+    } else {
+        try!(ui.warn(
+            "Unable to determine the type of app in your current directory",
+        ));
+        try!(ui.para(
+            "For now, we'll generate a plan with all of the available plan options and callbacks. \
+            For more documentation on plan options try passing --withdocs or visit \
+            https://www.habitat.sh/docs/reference/plan-syntax/",
+        ));
+        Ok(None)
+    }
+}
+
+fn is_project_ruby(path: &Path) -> bool {
```

Comment:
When you're dealing with filepaths as a function argument it's best to use a generic representation instead of a ref to `Path` or `PathBuf`. Here is an example:

```
fn is_project_ruby<T>(path: T> -> bool where T: AsRef<Path> {
    if path.as_ref().join("Gemfile).is_file() {
        return true;
    }
    // ...
}
```

I would recommend changing the functions in this file to leverage this approach. It makes the function easier to deal with for other developers. With this representation you can pass any thing which implements `AsRef<Path>` which includes, `str`, `&str`, `String`, `&String`, `Path`, `&Path`, `PathBuf`, `&PathBuf`, `OsStr`, `&OsStr`, and more